client/services/LocalStorageService.js

```diff
@@ -0,0 +1,196 @@
+// limitations under the License.
+/**
+ * @fileoverview A service that saves student code to the browser's
+ *    localStorage.
+ */
+
+tie.factory('LocalStorageService', ['FeedbackParagraphObjectFactory',
+  'ReinforcementBulletObjectFactory',
+  function(FeedbackParagraphObjectFactory, ReinforcementBulletObjectFactory) {
+    // In some browsers, localStorage is not available and its
+    // invocation throws an error.
+    var localStorageIsAvailable = false;
+    try {
+      localStorageIsAvailable = Boolean(localStorage);
+    } catch (e) {
+      localStorageIsAvailable = false;
+    }
+
+    /**
+     * Returns the local storage key for a given question.
+     *
+     * @param {string} questionId
+     * @param {string} language
+     * @returns {string}
+     */
+    var getLocalStorageKeyForCode = function(questionId, language) {
+      return questionId + ":" + language + ":code";
+    };
+
+    /**
+     * Returns the local storage key for the feedback.
+     *
+     * @param {string} questionId
+     * @param {string} language
+     * @returns {string}
+     */
+    var getLocalStorageKeyForFeedback = function(questionId, language) {
+      return questionId + ":" + language + ":feedback";
+    };
+
+    return {
+      /**
+       * Checks if the local storage is available.
+       *
+       * @returns {boolean}
+       */
+      isAvailable: function() {
+        return localStorageIsAvailable;
+      },
+
+      /**
+       * Stores the code into local storage - so long as local storage is
+       * available.
+       *
+       * @param {string} questionId
+       * @param {string} code
+       * @param {string} language
+       */
+      storeCode: function(questionId, code, language) {
+        if (!localStorageIsAvailable) {
+          return;
+        }
+
+        // Add version to the code, currently only 1
+        // TODO(talee): Add in check for the correct version number
+        var codeWithVersion = "1:" + code;
+
+        var localStorageKey = getLocalStorageKeyForCode(
+          questionId, language);
+        localStorage.setItem(localStorageKey, codeWithVersion);
+      },
+
+      /**
+       * Loads the code from local storage if local storage is available.
+       * If local storage is not available, then it returns null.
+       *
+       * @param {string} questionId
+       * @param {string} language
+       * @returns {string|null}
+       */
+      loadStoredCode: function(questionId, language) {
+        if (!localStorageIsAvailable) {
+          return null;
+        }
+
+        var localStorageKey = getLocalStorageKeyForCode(
+          questionId, language);
+        var storedCode = localStorage.getItem(localStorageKey);
+
+        if (storedCode === null) {
+          return null;
+        }
+        // TODO(talee): Check versioning when we start having different versions
+        storedCode = storedCode.substring(storedCode.indexOf(':') + 1);
```

Comment:
Don't redefine variables; create a new one for this. Maybe you could call the previous one "var storedCodeWithVersion".